### PR TITLE
Corrections to the game

### PR DESCRIPTION
**Pull request by Ajver**
# Added:

1. Wine is possible to buy only if that have sense (if elf has hp < max_hp)
1. Updating hp to max hp after buy vitality, if hp bar was full
1. Add small information popup near mouse (used for explaing items in shop, but in future may be used in other places as well)

### CODE REVIEW

File: Scenes/Tavern/PublicanScreen.gd [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	
+func _on_VinenBtn_bought() -> void:
+	vine_btn.call_deferred("set_enabled", false)
+	
+func should_wine_be_access() -> bool:
+	var elf = get_node("/root/World").find_node("Elf")
+	var elf_stats = get_node("/root/World/ElfStats")
+	var game_data = get_node("/root/World/GameData")
+	
+	return elf.hp < elf_stats.get_stat_value("vitality") and vine_btn.price <= game_data.gold
```
Comment (KazikBozia) [inline Scenes/Tavern/PublicanScreen.gd]: This line is too long. Maybe:
`var current_hp: int = elf_stats.get_stat_value("vitality")`
`return elf.hp < current_hp and vine_btn.price <= game_data.gold`

File: Resources/ElfStats.gd [hunk 2]
```diff
@@ -31,6 +31,25 @@ class Stat:
 		
 	func get_value() -> float:
 		return changed_value
+		
+	func get_value_replaced_item(item) -> float:
+		var vwc = value
+		
+		for c in changers:
+			if c.item_name != item.name:
+				vwc = c.get_multiplayed_value(vwc)
```
Comment (KazikBozia) [inline Resources/ElfStats.gd]: What is `c`? and `vwc`? maybe some better name or comment ?

File: Scenes/InformationPopup/Popup.tscn [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+margin_right = -6.0
+margin_bottom = -6.0
+grow_vertical = 2
+
+[node name="Label" type="Label" parent="MarginContainer"]
+margin_right = 55.0
+margin_bottom = 14.0
+grow_horizontal = 2
+grow_vertical = 2
+text = "addwa"
```
Comment (KazikBozia) [inline Scenes/InformationPopup/Popup.tscn]: :P
Maybe `default Popup text`

File: Scenes/InformationPopup/Popup.gd [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+export(float) var mouse_y_offset
+
+var mouse_pos : Vector2
+
+onready var label = find_node("Label")
+	
+func init(Title:String):
+	set_title(Title)
+	call_deferred("update_size_and_pos")
```
Comment (KazikBozia) [inline Scenes/InformationPopup/Popup.gd]: This line is useless

File: Scenes/InformationPopup/Popup.gd [hunk 5]
```diff
@@ -0,0 +1,42 @@
+extends Control
+
+export(float) var mouse_y_offset
+
+var mouse_pos : Vector2
+
+onready var label = find_node("Label")
+	
+func init(Title:String):
```
Comment (KazikBozia) [inline Scenes/InformationPopup/Popup.gd]: top-down design rule...
`init`, `set_title`, `update_size_and_pos`, `update_size`, `update_position` ---> low and lower levels of abstraction

File: Scenes/Tavern/PublicanScreen.gd [hunk 6]
```diff
@@ -0,0 +1,24 @@
+extends "res://Scenes/Tavern/OneTavernScreen.gd"
+
+onready var vine_btn = find_node("VineBtn")
+
+func _ready():
+	vine_btn.connect("bought", self, "_on_VinenBtn_bought")
+	
+func on_enter():
```
[the remark this answers is not in the document]
Comment (Ajver) [reply Scenes/Tavern/PublicanScreen.gd]: Right, that's good rule.

File: Scenes/Tavern/PublicanScreen.gd (hunk 1 above)
Comment (Ajver) [reply Scenes/Tavern/PublicanScreen.gd]: Good idea. But not current_hp, but max_hp ;)

File: Resources/ElfStats.gd (hunk 2 above)
Comment (Ajver) [reply Resources/ElfStats.gd]: Fixed:

c - changer,

File: Scenes/InformationPopup/Popup.tscn (hunk 3 above)
Comment (Ajver) [reply Scenes/InformationPopup/Popup.tscn]: Haha, I forgot about that placeholder ;D

File: Scenes/InformationPopup/Popup.gd (hunk 4 above)
Comment (Ajver) [reply Scenes/InformationPopup/Popup.gd]: Well, yes.

File: Scenes/InformationPopup/Popup.gd (hunk 5 above)
Comment (Ajver) [reply Scenes/InformationPopup/Popup.gd]: Right. I moved `_input` function above the `update_size`. Do You think it's OK?

File: Scenes/Tavern/PublicanScreen.gd (hunk 1 above)
Comment (KazikBozia) [reply Scenes/Tavern/PublicanScreen.gd]: Ok, actually

**Pull request merged**